oauth2/oauth2-jose/src/main/java/org/springframework/security/oauth2/jwt/NimbusJwtDecoderJwkSupport.java

```diff
@@ -99,13 +105,57 @@ public NimbusJwtDecoderJwkSupport(String jwkSetUrl, String jwsAlgorithm) {
 		} catch (MalformedURLException ex) {
 			throw new IllegalArgumentException("Invalid JWK Set URL \"" + jwkSetUrl + "\" : " + ex.getMessage(), ex);
 		}
+
 		this.jwsAlgorithm = JWSAlgorithm.parse(jwsAlgorithm);
 		JWSKeySelector<SecurityContext> jwsKeySelector =
 			new JWSVerificationKeySelector<>(this.jwsAlgorithm, jwkSource);
+
 		this.jwtProcessor = new DefaultJWTProcessor<>();
 		this.jwtProcessor.setJWSKeySelector(jwsKeySelector);
 	}
 
+	/**
+	 * Constructs a {@code NimbusJwtDecoderJwkSupport} using the provided parameters.
+	 *
+	 * Note that this decoder will first verify that the "alg" claim in the JWT matches
+	 * the configured algorithm before verifying the signature.
+	 *
+	 * @param key - the public key to use when verifying the signature
+	 *
+	 * @since 5.1
+	 */
+	public NimbusJwtDecoderJwkSupport(RSAPublicKey key) {
+		this(key, JwsAlgorithms.RS256);
+	}
+
+	/**
+	 * Constructs a {@code NimbusJwtDecoderJwkSupport} using the provided parameters.
+	 *
+	 * Note that this decoder will first verify that the "alg" claim in the JWT matches
+	 * the configured algorithm before verifying the signature.
+	 *
+	 * @param key - the public key to use when verifying the signature
+	 * @param jwsAlgorithm - the expected algorithm
+	 *
+	 * @since 5.1
+	 */
+	public NimbusJwtDecoderJwkSupport(PublicKey key, String jwsAlgorithm) {
```

Comment:
In case this seems odd, I'll clarify here that this constructor is in this class because it correlates to a constructor in NimbusReactiveJwtDecoder, which supports both remote and local key configuration. The name NimbusJwtDecoderJwkSupport evokes only remote support, which may cause confusion.

I'm actually partial to keeping the two use cases separate, mostly because part of their state is mutually exclusive. For example, the remote construction of this class needs a RestOperations, whereas the local construction does not. If this is instantiated with a public key, then setRestOperations becomes a no-op which may be confusing to users. It makes the class feel less cohesive when it supports both uses cases.

Reply:
> the remote construction of this class needs a RestOperations, whereas the local construction does not. If this is instantiated with a public key, then setRestOperations becomes a no-op which may be confusing to users

I totally agree. We should have a separate implementation of `JwtDecoder` for RSA key configuration.

Reply:
Let's see what @rwinch thinks as this would likely have reactive implications.

Reply:
I really think that we should just expose this as the same API with different builders as static methods on the class. The logic is the same, but how you build the implementation just differs.

Reply:
I will add that the name NimbusJwtDecoderJwkSupport is misleading. I find this quite unfortunate. Perhaps we should create a NimbusJwtDecoder that is more generic and supports the builder. The  NimbusJwtDecoderJwkSupport would be deprecated and just delegate to the NimbusJwtDecoder

Reply:
I think my concerns would be allayed if there weren't state hanging around that was sometimes used in one circumstance but not another, e.g. rest operations (or the jwkSetUrl before that).

For example, if this class had only an immutable instance of the Nimbus JWTProcessor, I think keeping everything in the class makes more sense since how the key is obtained shouldn't matter to this class.

If we do something like this (https://github.com/spring-projects/spring-security/issues/5648), for example, say down the road, then I don't think we need two separate classes. (Not saying we need to specifically do what that tasks says here, it just an example of what I'm thinking about.)

Reply:
Let's take this in stride then. Please first create a ticket for NimbusJwtDecoder which follows this approach. We can add builders for supporting the various use cases. Then NimbusJwtDecoderJwkSupport will internally delegate to NimbusJwtDecoder